rust/src/mqtt/mqtt.rs

```diff
@@ -0,0 +1,777 @@
+                        let mut tx = self.new_tx(msg, toclient);
+                        &mut MQTTState::set_event(&mut tx, MQTTEvent::InvalidQosLevel);
+                        self.transactions.push(tx);
+                    }
+                }
+            },
+            MQTTOperation::SUBSCRIBE(ref subscribe) => {
+                if !self.connected {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::UnintroducedMessage);
+                    self.transactions.push(tx);
+                    return;
+                }
+                let pkt_id = subscribe.message_id as u32;
+                match msg.header.qos_level {
+                    0 => {
+                        // with QOS level 0, we do not need to wait for a
+                        // response
+                        let mut tx = self.new_tx(msg, toclient);
+                        tx.complete = true;
+                        self.transactions.push(tx);
+                    },
+                    1..=2 => {
+                        let mut tx = self.new_tx(msg, toclient);
+                        tx.pkt_id = Some(pkt_id);
+                        self.transactions.push(tx);
+                    },
+                    _ => {
+                        let mut tx = self.new_tx(msg, toclient);
+                        &mut MQTTState::set_event(&mut tx, MQTTEvent::InvalidQosLevel);
+                        self.transactions.push(tx);
+                    }
+                }
+            },
+            MQTTOperation::UNSUBSCRIBE(ref unsubscribe) => {
+                if !self.connected {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::UnintroducedMessage);
+                    self.transactions.push(tx);
+                    return;
+                }
+                let pkt_id = unsubscribe.message_id as u32;
+                match msg.header.qos_level {
+                    0 => {
+                        // with QOS level 0, we do not need to wait for a
+                        // response
+                        let mut tx = self.new_tx(msg, toclient);
+                        tx.complete = true;
+                        self.transactions.push(tx);
+                    },
+                    1..=2 => {
+                        let mut tx = self.new_tx(msg, toclient);
+                        tx.pkt_id = Some(pkt_id);
+                        self.transactions.push(tx);
+                    },
+                    _ => {
+                        let mut tx = self.new_tx(msg, toclient);
+                        &mut MQTTState::set_event(&mut tx, MQTTEvent::InvalidQosLevel);
+                        self.transactions.push(tx);
+                    }
+                }
+            },
+            MQTTOperation::CONNACK(ref _connack) => {
+                if let Some(tx) = self.get_tx_by_pkt_id(MQTT_CONNECT_PKT_ID) {
```

Comment:
Is there no risk of collision against `MQTT_CONNECT_PKT_ID` ?

Reply:
There is no risk. Message IDs (also called 'packet identifiers' in some spec versions) in MQTT messages are only 16 bit, so no ID parsed from traffic can collide with that special value `MQTT_CONNECT_PKT_ID` (which is defined to be `std::u32::MAX`).

Reply:
Thanks for the explanation. 
Can you add this as a comment ?

Reply:
Done.